### PR TITLE
Tail call optimization

### PR DESCRIPTION
**Pull request by james-perretta**

### CODE REVIEW

File: find_tail_calls.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    def visit_SFunction(self, func: sexp.SFunction) -> None:
+        if func.is_lambda:
+            # We're only looking for statically-detectable tail recursion.
+            return
+
+        body_exprs = list(func.body)
+        assert len(body_exprs) > 0, 'Function bodies must not be empty'
+
+        for body_expr in body_exprs[:-1]:
```
Comment (porglezomp) [inline find_tail_calls.py]: This is bailing out if there are any non-tail recursive calls, right?

I believe:
```scheme
(define (M n)
  (if (> n 100)
     (- n 10)
     (M (M (+ n 11)))))
```
can be safely optimized to
```
M(n) =
  while n <= 100
    n := M(n + 11)
  n - 10
```

File: find_tail_calls.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    def visit_SCall(self, call: sexp.SCall) -> None:
+        if self._current_function is None:
+            return
+
+        if (isinstance(call.func, sexp.SSym)
+                and call.func != self._current_function.name):
+            return
+
+        recursive_call_finder = RecursiveCallFinder(self._current_function)
+        for arg_expr in call.args:
```
Comment (porglezomp) [inline find_tail_calls.py]: Similarly, I don't think we need to worry about this case, either, it's still tail recursive even if it makes another tail recursive call.

File: emit_IR.py [hunk 3]
```diff
@@ -49,6 +58,8 @@ def visit_SFunction(self, func: sexp.SFunction) -> None:
         )
 
         func.code = emitted_func
+        assert func.name not in self.global_env, (
+            f"Duplicate function name: {func.name}")
```
Comment (porglezomp) [inline emit_IR.py:62]: what, you don't like shadowing? :P
(I realize that with our current dynamic binding, shadowing makes the recursion optimization unsound…)

Comment (porglezomp) [reply emit_IR.py:62]: Ah, we could avoid this if we made the function's name always available as a local? Then it would be lexically scoped, and also would avoid a `lookup` instruction.

Comment (porglezomp) [reply emit_IR.py:62]: Anyway, this is just me thinking, the `assert` is fine and we'll leave it as is.

**Review by porglezomp — CHANGES_REQUESTED** [review]

File: test_builtins.py [hunk 4]
```diff
@@ -64,7 +67,11 @@ def test_builtins(self) -> None:
         self.assertEqual(run(env, '(symbol? 42)'), SBool(False))
         self.assertEqual(run(env, '(vector? [])'), SBool(True))
         self.assertEqual(run(env, '(vector? 42)'), SBool(False))
-        self.assertEqual(run(env, '(function? (lambda () []))'), SBool(True))
+        # Use a copy of the environment, since (lambda) adds a name to
```
Comment (porglezomp) [inline test_builtins.py:70]: Why is this necessary?

File: test_find_tail_calls.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        self.assert_symbol_in_tail_calls(
+            sexp.SSym('vacuous-tail'), self.finder.tail_calls)
+
+    def test_recursive_call_not_last_expr(self) -> None:
+        prog = sexp.parse(
+            '(define (not-tail) (not-tail) (if true true false) (not-tail))')
+        self.finder.visit(prog)
+        self.assertEqual(0, len(self.finder.tail_calls))
+
+    def test_potential_tail_call_has_recusive_call_as_args(self) -> None:
```
Comment (porglezomp) [inline test_find_tail_calls.py]: The outer call is still a tail call

File: test_find_tail_calls.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        prog = sexp.parse('''
+            (define (vacuous-tail)
+                (if true (+ 1 (vacuous-tail)) (vacuous-tail))
+            )''')
+        self.finder.visit(prog)
+        self.assertEqual(1, len(self.finder.tail_calls))
+        self.assert_symbol_in_tail_calls(
+            sexp.SSym('vacuous-tail'), self.finder.tail_calls)
+
+    def test_recursive_call_not_last_expr(self) -> None:
```
Comment (porglezomp) [inline test_find_tail_calls.py:55]: Still can do TCO

File: test_builtins.py (hunk 4 above)
Comment (james-perretta) [reply test_builtins.py:70]: If you `run` two programs that define lambdas with the same environment, you'll get a duplicate lambda name. Copying the env whenever that happens makes it robust against future changes that introduce lambdas to the environment.

File: find_tail_calls.py (hunk 1 above)
Comment (james-perretta) [reply find_tail_calls.py]: Yes, I did that on purpose, but I can change it if you think there's a significant benefit to doing that transformation

File: test_builtins.py (hunk 4 above)
Comment (porglezomp) [reply test_builtins.py:70]: Ah… I expected the lambda names to be globally unique, to my mind it should be perfectly safe to run multiple times.

**Review by porglezomp — APPROVED** [review]

File: test_builtins.py (hunk 4 above)
Comment (james-perretta) [reply test_builtins.py:70]: Yeah, I considered making that the case, but then the test cases would start depending on the order they're run in...
